### PR TITLE
feat: implement issue #686 — canary #668 increment 5: Layer 0 — pure decision cores + codified 'decision logic in a tested script' standard

### LINKED ISSUE #686: canary #668 increment 5: Layer 0 — pure decision cores + codified 'decision logic in a tested script' standard
Increment 5 of #668 (**Layer 0** of the [design](https://github.com/petry-projects/.github/issues/668#issuecomment-4949644993)) — **pure, unit-tested decision cores + the codified standard**. This is where most correctness bugs get caught **before a version exists** — cheaper and broader than any canary-time signal.

## Context / what's already done
#680/#682 set the precedent: criterion #2 of pr-auto-review's readiness gate (all **required** checks passing) was extracted into `.github/scripts/pr-auto-review/lib/ready-check.sh` (`pr_auto_review_checks_ready`), with bats in `test/workflows/pr-auto-review/` gated by the existing **PR Auto-Review Tests** workflow. Increment 5 finishes the job for pr-auto-review and writes the standard.

### 1. Extract the remaining readiness criteria into the pure lib
Today the `check-and-dispatch` job in `.github/workflows/pr-auto-review-reusable.yml` still has criteria **#1** (PR open + not draft), **#3** (effective review decision ≠ CHANGES_REQUESTED), and **#4** (no unresolved review threads) as **inline bash**. Move their decision logic into pure functions in `ready-check.sh` (e.g. `pr_auto_review_ready <state> <is_draft> <checks_json> <required_json> <self_check> <review_decision> <unresolved_count>` returning the ready/skip verdict **and** the decision class), so the workflow only does I/O (gather PR state, checks, review decision, unresolved-thread count via gh/GraphQL) and calls the pure core.
- **Preserve the `decision=<class>` outputs** the job emits (`skip-draft`, `skip-checks-pending`, `skip-changes-requested`, `skip-unresolved-threads`, `dispatched`) — Layer 2 decision-telemetry (increment 4) consumes them. The pure function should RETURN the class; the glue echoes it to `$GITHUB_OUTPUT`.
- Keep `pr_auto_review_checks_ready` (from #682) as-is or fold it into the unified core — no behaviour change to the required-checks gate.

### 2. Decision-matrix bats (`test/workflows/pr-auto-review/`)
Cover the full matrix, one point per branch, plus combinations that exercise precedence (e.g. draft + CHANGES_REQUESTED → whichever the gate checks first):
- open+non-draft, all-required-green, no CHANGES_REQUESTED, 0 unresolved → **ready=true / dispatched**
- draft → skip-draft; closed → not-ready
- a required check pending/failing → skip-checks-pending (a NON-required failing/cancelled check must NOT block — the #680 fix; pin it)
- CHANGES_REQUESTED → skip-changes-requested
- ≥1 unresolved thread → skip-unresolved-threads
- no checks at all → skip-checks-pending

### 3. Codify the standard
Add to `AGENTS.md` (and/or `standards/agent-standards.md`) a short standard: **"A decision-making reusable keeps its decision logic in a pure, side-effect-free, unit-tested script (`scripts/**`), with the workflow as thin I/O glue. Gate the script with bats in CI on any PR that changes the reusable."** Cite the canary engine (`scripts/lib/canary-rollout.sh` + `tests/canary_rollout.bats`, now CI-gated via #685) and pr-auto-review's `ready-check.sh` as the exemplars. Note the payoff: correctness bugs caught pre-merge, whole decision matrix covered vs one fixture point, zero canary-time cost.

## Out of scope
- Extending the pattern to other agents' reusables (auto-rebase already follows it; dependabot-automerge/dev-lead are follow-ups, not this issue).
- Increment 6 (optional SUSPECT→PRE_EXISTING auto-downgrade) — deferred per design until SUSPECT toil appears.

## Acceptance
- Criteria #1/#3/#4 decision logic lives in `ready-check.sh` pure functions; `check-and-dispatch` is I/O glue that calls them.
- `decision=<class>` outputs unchanged (Layer 2 telemetry still works); required-vs-non-required gate (#680) behaviour unchanged.
- Full decision-matrix bats green under the **PR Auto-Review Tests** workflow.
- Standard codified in AGENTS.md / agent-standards.md with the two exemplars.
- shellcheck clean.

### PR DESCRIPTION
**Pull request by don-petry**
Closes #686

Implemented by dev-lead agent. Please review.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved pull request review readiness decisions with consistent handling for drafts, closed or merged requests, required checks, requested changes, and unresolved threads.
  * Non-required check failures no longer block review dispatch.
  * Added clear decision outcomes and consistent precedence when multiple conditions apply.

* **Documentation**
  * Documented the review readiness criteria and decision flow.
  * Added standards for keeping reusable decision logic independently tested and workflow integration lightweight.

* **Tests**
  * Added comprehensive coverage for readiness scenarios and edge cases.

### CODE REVIEW

**Comment by don-petry** [comment]
## Dev-Lead — review-changes (applied)
Changes committed and pushed.

File: .github/scripts/pr-auto-review/lib/ready-check.sh [hunk 1]
```diff
@@ -93,3 +93,69 @@ pr_auto_review_checks_ready() {
...
+  #    delegated to pr_auto_review_checks_ready (issue #680, unchanged).
+  [ -z "$checks_json" ] && checks_json="[]"
+  local total
+  total=$(printf '%s' "$checks_json" | jq 'length')
+  if [ "$total" -eq 0 ] \
+     || ! printf '%s' "$checks_json" \
+          | pr_auto_review_checks_ready "$required_json" "$self_name" >/dev/null; then
+    echo "skip-checks-pending"
+    return 1
+  fi
```
[the remark this answers is not in the document]
Comment (don-petry) [reply .github/scripts/pr-auto-review/lib/ready-check.sh:143]: Fixed in `.github/scripts/pr-auto-review/lib/ready-check.sh`: replaced `jq 'length'` with `jq 'if type == "array" then length else 0 end' 2>/dev/null` and added a `[ -z "$total" ]` guard before the integer comparison. The old normalization line (`[ -z "$checks_json" ] && checks_json="[]"`) is superseded by the inline type check, which safely returns 0 for JSON null, non-array values, or completely malformed input — preventing the unary-operator Bash error.

File: test/workflows/pr-auto-review/ready.bats [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+  run pr_auto_review_ready "OPEN" "false" '[]' "$REQUIRED" "" "APPROVED" "0"
+  [ "$status" -eq 1 ]
+  [ "$output" = "skip-checks-pending" ]
+}
+
+@test "not ready: empty checks string normalized → skip-checks-pending" {
+  run pr_auto_review_ready "OPEN" "false" "" "$REQUIRED" "" "APPROVED" "0"
+  [ "$status" -eq 1 ]
+  [ "$output" = "skip-checks-pending" ]
+}
```
[the remark this answers is not in the document]
Comment (don-petry) [reply test/workflows/pr-auto-review/ready.bats:98]: Added in `test/workflows/pr-auto-review/ready.bats`: new test `"not ready: null or malformed checks string → skip-checks-pending"` that asserts both `"null"` and `"invalid-json"` inputs produce `skip-checks-pending` (exit 1). The test passes against the updated implementation in `ready-check.sh` and all 19 tests in the suite are green.

**Comment by don-petry** [comment]
## Dev-Lead — review-changes (applied)
Changes committed and pushed.

File: .github/scripts/pr-auto-review/lib/ready-check.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (don-petry) [reply .github/scripts/pr-auto-review/lib/ready-check.sh:143]: Verified: the fix is applied in `.github/scripts/pr-auto-review/lib/ready-check.sh` (line 137–143). `jq 'length'` has been replaced with `jq 'if type == "array" then length else 0 end' 2>/dev/null` and the `[ -z "$total" ]` guard precedes the integer comparison, preventing the unary-operator Bash error on null/non-array/malformed input. All 19 tests in the ready.bats suite pass, including the new null/malformed coverage.

File: test/workflows/pr-auto-review/ready.bats (hunk 2 above)
[the remark this answers is not in the document]
Comment (don-petry) [reply test/workflows/pr-auto-review/ready.bats:98]: Verified: the test is added in `test/workflows/pr-auto-review/ready.bats` (lines 100–108) as `"not ready: null or malformed checks string → skip-checks-pending"`. It asserts both `"null"` and `"invalid-json"` inputs return exit 1 with output `skip-checks-pending`. All 19 tests in the suite are green.

**Comment by don-petry** [comment]
## Dev-Lead — fix-reviews (applied)
Changes committed and pushed.

**Comment by don-petry** [comment]
## Dev-Lead — review-changes (no-changes)
No changes were needed for this PR.

File: AGENTS.md [hunk 3]
```diff
@@ -111,6 +111,24 @@ If a dependency cannot be resolved, report the specific blocker and a workaround
...
+- **Canary rollout engine** — `scripts/lib/canary-rollout.sh` (pure gate core) + `tests/canary_rollout.bats`, CI-gated by `canary-rollout-tests.yml` (#685).
+  The orchestrator `scripts/canary-rollout.sh` feeds it numbers gathered from `gh`.
+- **PR auto-review readiness gate** — `.github/scripts/pr-auto-review/lib/ready-check.sh` (`pr_auto_review_ready` + helpers) + `test/workflows/pr-auto-review/`, CI-gated by
+  the **PR Auto-Review Tests** workflow. The reusable `pr-auto-review-reusable.yml` is thin glue that gathers PR state and calls the pure core.
+
+Why it pays off: correctness bugs are caught **pre-merge** by fast unit tests instead of at canary time or in production;
+the **whole decision matrix** is covered (every branch, precedence, edge cases) rather than the single fixture point a live run exercises;
+and it adds **zero canary-time cost** because the logic is verified before a version ever ships.
+Inline bash trapped in YAML can be exercised only by triggering the workflow — extract it.
+
```
[the remark this answers is not in the document]
Comment (don-petry) [reply AGENTS.md:131]: Fixed in `AGENTS.md` line 116: updated the glob pattern from `scripts/**` to `scripts/**` or `.github/scripts/**` (using `e.g.` phrasing per the proposed wording) so it no longer contradicts the PR auto-review exemplar at `.github/scripts/pr-auto-review/lib/ready-check.sh`. The duplication concern is addressed separately on the `standards/agent-standards.md` thread.

File: standards/agent-standards.md [hunk 4]
```diff
@@ -109,6 +109,29 @@ For repos with `package.json` referencing BMAD modules (e.g., `bmad-method`,
...
+| Canary rollout engine | `scripts/lib/canary-rollout.sh` | `tests/canary_rollout.bats` | `canary-rollout-tests.yml` (#685) |
+| PR auto-review readiness | `.github/scripts/pr-auto-review/lib/ready-check.sh` | `test/workflows/pr-auto-review/` | **PR Auto-Review Tests** |
+
+**Payoff:** correctness bugs are caught pre-merge by fast unit tests instead of
+at run time; the whole decision matrix is covered (every branch and precedence
+edge) rather than the single fixture point a live trigger exercises; and it adds
+zero run-time cost because the logic is verified before the workflow ever fires.
+Decision logic trapped as inline bash in YAML can only be exercised by triggering
+the workflow — extract it into a tested script.
+
```
[the remark this answers is not in the document]
Comment (don-petry) [reply standards/agent-standards.md:116]: Fixed in `standards/agent-standards.md`: replaced the full duplicated "Decision-Making Reusables" section (rules + exemplar table + payoff text) with a concise cross-reference to `AGENTS.md § Decision Logic Lives in a Pure, Tested Script`, following the same convention as the adjacent "BMAD Method Workflows" section which links to ci-standards.md rather than duplicating it.

**Comment by don-petry** [comment]
## Dev-Lead — fix-reviews (applied)
Changes committed and pushed.

**Comment by don-petry** [comment]
## Dev-Lead — waiting on PR blockers (intent: fix-reviews)
**PR:** #687
No changes were committed, but the PR still has blocking checks or reviews (failing or cancelled checks, or changes-requested reviews). The retry cron will re-attempt automatically. Next attempt after: `2026-07-12T23:57:07Z`
